presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -88,401 +100,465 @@
+        catch (Exception e) {
+            logger.warn("Failed to rewrite query with materialized view with following exception: %s", e.getMessage());
             return node;
         }
     }
 
     @Override
-    protected Node visitNode(Node node, Void context)
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
     {
-        for (Node child : node.getChildren()) {
-            process(child, context);
+        if (!node.getFrom().isPresent()) {
+            throw new IllegalStateException("Query with no From clause is not rewritable by materialized view");
+        }
+        // If a query specification has a Table as a source, it can potentially be rewritten,
+        // so hand control over to LeafSubqueryVisitor via rewriteQuerySpecificationIfCompatible
+        Table baseTable;
+        if (node.getFrom().get() instanceof Table) {
+            baseTable = (Table) node.getFrom().get();
+            return rewriteQuerySpecificationIfCompatible(node, baseTable);
+        }
+        else if (node.getFrom().get() instanceof AliasedRelation
+                        && ((AliasedRelation) node.getFrom().get()).getRelation() instanceof Table) {
+            AliasedRelation aliasedRelation = (AliasedRelation) node.getFrom().get();
+            baseTable = (Table) aliasedRelation.getRelation();
+            return rewriteQuerySpecificationIfCompatible(node, baseTable);
+        }
+        else {
+            return super.visitQuerySpecification(node, context);
         }
-        return node;
     }
 
-    @Override
-    protected Node visitQuery(Query node, Void context)
+    private QuerySpecification rewriteQuerySpecificationIfCompatible(QuerySpecification querySpecification, Table baseTable)
     {
-        return new Query(
-                node.getWith(),
-                (QueryBody) process(node.getQueryBody(), context),
-                node.getOrderBy(),
-                node.getOffset(),
-                node.getLimit());
-    }
+        QualifiedObjectName tableName = createQualifiedObjectName(session, baseTable, baseTable.getName());
+        List<ConnectorMaterializedViewDefinition> linkedMaterializedViews = baseTableToMaterializedViewMap.getOrDefault(tableName, ImmutableList.of());
 
-    @Override
-    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+        for (ConnectorMaterializedViewDefinition materializedViewDefinition : linkedMaterializedViews) {
+            Table materializedViewTable = new Table(QualifiedName.of(materializedViewDefinition.getTable()));
+            Query materializedViewQuery = (Query) sqlParser.createStatement(materializedViewDefinition.getOriginalSql(), createParsingOptions(session));
+
+            QuerySpecification rewritten = new LeafSubqueryVisitor(materializedViewTable, materializedViewQuery).rewrite(querySpecification);
+
+            if (rewritten != querySpecification) {
```

Comment:
should we use !equals() here?

Reply:
I had the same thought - this was how it was being done before and it made me curious. I honestly don't know why this comparison works (it *does* do an equality based comparison, not identity like you'd expect). Searching seems to show that `!=` should do identity based comparison. Likewise, this:

```
        Map map = Map.of("abc", "def");

        Map map2 = Map.of("abc", "def");

        System.out.println(map != map2);
```

returns `true`. My best guess is we intern the object somewhere or something, but I don't know much about that. I'd be curious to hear from someone who knows.